internal/restorer/filerestorer.go

```diff
@@ -101,6 +102,7 @@ func (r *fileRestorer) forEachBlob(blobIDs []restic.ID, fn func(packID restic.ID
 func (r *fileRestorer) restoreFiles(ctx context.Context) error {
 
 	packs := make(map[restic.ID]*packInfo) // all packs
+	var packOrder restic.IDs
```

Comment:
The intuition behind the `packOrder` optimization is that any order, even if it only helps at a few places, provides better locality than the current completely random restore order. At least it shouldn't be worse on average.

Regarding the out-of-order writes: The operating system usually caches writes for a short time, which essentially allows it to properly order nearby out-of-order writes. Together with preallocated files this has the effect, that longer stretches of a file can be written sequentially. Note that the page cache also helps to ensure that the out-of-order writes are written in an optimized order, but this will still mean seeking around in the file after each block.

Without preallocation the exact effect of `packOrder` depends a lot on the filesystem implementation. For example the kernel wiki states for ext4 that (from https://ext4.wiki.kernel.org/index.php/Ext4_Disk_Layout#Block_and_Inode_Allocation_Policy):
>  A second related trick that ext4 uses is delayed allocation. Under this scheme, when a file needs more blocks to absorb file writes, the filesystem defers deciding the exact placement on the disk until all the dirty buffers are being written out to disk. By not committing to a particular placement until it's absolutely necessary (the commit timeout is hit, or sync() is called, or the kernel runs out of memory), the hope is that the filesystem can make better location decisions.

The restorer doesn't call sync(), so if nearby blocks are written in short succession in any order, the actual data placement will happen later on, thus properly ordering the out-of-order writes on disk, which at least yields larger fragments. Fullly random writes within a file are quite unlikely to close most gaps within the timeout.

On macOS apfs (without preallocation) seems to be happy to scatter fileparts just about everywhere even if a file is written linearly, so it's hard to tell of how much use the optimization will be there, but I guess it doesn't matter much on SSDs. On Windows on the other hand, files are by default not sparse and thus writing at a random place will force the operation system to zero blocks up to the write location. Writing files sequentially will help avoiding to write large parts of a file twice (assuming the file is too large for the page cache or the restorer takes too much time to fill in the gaps).

So I hope this alleviates the second concern.

Regarding the first concern, I'll try to come up with a metric how to measure this. My expectation would be that `packOrder` becomes less effective the older a repository is, as blocks are probably scattered around much more.

Reply:
Another idea how to demonstrate `packOrder` effectiveness is to purposely change implementation to produce worst possible pack order, then compare how restore performs using "good" and "bad" `packOrder`.

Reply:
I've hacked together an experiment branch https://github.com/MichaelEischer/restic/tree/preallocate-experiment . The modifications disable all parts of the restorer which actually load / write data. I'm not too sure whether the metric I've come up with is that useful. It tries to count how often restore has to jump between file parts / files.

So I hope it captures the cause of file fragmentation. In regard to restore speed there shouldn't be much difference on SSDs, but I expect that less "jumps" translate to less seeking on HDDs. However, it all depends a lot on the underlying filesystem (cache).

The metric only considers files which are large enough to consist of multiple blobs (For small files, we can't optimize the write order, so just ignore them to reduce noise.). For multi-blob files it counts how often writing the next blob would require jumping inside a file or between files. To represent a crude filesystem cache a write counts as sequential (i.e. not a jump) when it is adjacent to one of the last 20 blobs. That is blobs written in reverse order (last to first) are not counted as jumps, as the filesystem cache should be able to merge most of these writes. The first access to a multi-blob file is also not counted.

While debugging the metric, I've made an interesting observation: During a backup consecutive chunks of a file usually are written one after the other into packfiles. With multiple goroutines and varying chunks sizes, these blobs are stored slightly out of order, such that we might see the chunks of a file packed into two pack files like `1 5 3 2` and `4 8 6 7` (number = i-th chunk in file). This slight disorder is hopefully smoothed out by the filesystem.

For my personal backup (a mix of documents, music, program code, VMs. 306GB, 535k files) it get the following results:
```
Jumps ordered 64351
Jumps random 166241
Jumps standard 166378
```
The `ordered` variant uses the order from `packOrder`, `random` uses a shuffled version of `packOrder` and `standard` is the current implementation. So the ordered variants seems to work better in that regard, however, it's hard to say what the exact effects are performance-wise.

My assumption is that writing file chunks in a random order should be the worst case from a performance and fragmentation perspective. Under that assumption, the _current_ restorer implementation already represents the worst case. We could "optimize" the order to get rid of a few randomly occurring cases of nearby accesses, but I guess doing that properly is NP-complete and the random order should do just fine.

I'm currently also experimenting with restoring some parts of my backup with with the current master, just preallocation and preallocation + ordered packs.

Reply:
> but preallocation solves file fragmentation problem

That assumes that preallocation always works, which is quite likely not always the case. Just using `packOrder` alone without preallocation has also helped a lot regarding fragmentation in my tests.

Regarding restore performance I've run a few tests. These were running on a Raspberry Pi 4 (4 GB RAM), the repository _and_ the restore destination are stored on a 8TB external HDD (`WDC WD80EZAZ-11TDBA0`). I've used the raspberry as it's directly connected to my backup disk. The user/sys time reported by the `time` command suggests that on average only 2-3 of the 4 cores are used, so the experiments seem to be IO-bound. With both the repository and the restore destination stored on the same disk this basically amounts to a worst-case scenario regarding reading and writing from/to random spots on disk.

I've used two dataset, one with disk images of a small virtual machine and the other one is a mix of documents, tons of small files and a few larger ones (primarily git-pack files). A test run covers all restorer variants and both datasets, run in a row. After each variant there's also a sleep of 2 minutes (to give the system some time to settle and maybe cool down a bit between test runs) and a call to `sysctl -w vm.drop_caches=3`.

Variants: `master` is the current master, `prealloc` just issues the preallocate calls and `order` combines both preallocation and ordered packs.

Debian 9 VM (16GB, 2 disk images)

variant | run 1 | run 2 | run 3 | run 4 | run 5
--- | --- | --- | --- | --- | ---
master | 14m35.819s | 14m2.406s | 15m10.766s | 14m13.133s | 15m13.717s
prealloc | 14m57.076s | 13m56.299s | 13m58.150s | 15m9.552s | 14m49.344s
order | 11m19.405s | 11m59.684s | 11m49.451s | 11m49.538s | 12m0.065s


Documents / Code / Git repos (18GB, 234k files)

variant | run 1 | run 2 | run 3 | run 4 | run 5
--- | --- | --- | --- | --- | ---
master | 19m52.188s | 20m0.253s |20m2.241s | 20m20.528s | 20m20.477s
prealloc | 19m32.632s | 22m33.424s | 22m18.826s | 22m17.385s | 22m47.300s
order | 18m19.280s | 18m24.466s | 18m27.536s | 18m20.145s | 18m26.163s

After massaging the results into the format expected by benchstat, I get the following results:
```
name                           old s       new s       delta
RestoreVM-master-vs-prealloc     879 ± 4%    874 ± 4%   ~     (p=0.548 n=5+5)
RestoreVM-master-vs-order        879 ± 4%    708 ± 4%  -19.51%  (p=0.008 n=5+5)

RestoreMix-master-vs-prealloc  1.21k ± 1%  1.31k ±11%   ~     (p=0.151 n=5+5)
RestoreMix-master-vs-order     1.21k ± 1%  1.10k ± 0%   -8.58%  (p=0.008 n=5+5)
```

For the VM repository the prealloc variant doesn't change anything regarding performance and for the second test repository, the prealloc variant might even be a bit slower.

But the clear trend here is that using `packOrder` provides a speed-up between 8 and 19%.

Reply:
This is convincing. Thank you for putting up with me. Can you add a comment to `packOrder` to explain that while it does not provide _guaranteed_ file restore order, the opportunistic order it does provide improved restore performance by 19% in some tests?

Reply:
I've added a short comment.